core/src/main/java/org/apache/calcite/rel/rules/AggregateUnionTransposeRule.java

```diff
@@ -125,30 +125,34 @@ public AggregateUnionTransposeRule(Class<? extends Aggregate> aggregateClass,
       return;
     }
 
-    // create corresponding aggregates on top of each union child
-    final RelBuilder relBuilder = call.builder();
-    int transformCount = 0;
+    boolean hasNonUniqueInputs = false;
     final RelMetadataQuery mq = call.getMetadataQuery();
     for (RelNode input : union.getInputs()) {
       boolean alreadyUnique =
           RelMdUtil.areColumnsDefinitelyUnique(mq, input,
               aggRel.getGroupSet());
 
-      relBuilder.push(input);
       if (!alreadyUnique) {
-        ++transformCount;
-        relBuilder.aggregate(relBuilder.groupKey(aggRel.getGroupSet()),
-            aggRel.getAggCallList());
+        hasNonUniqueInputs = true;
+        break;
       }
     }
 
-    if (transformCount == 0) {
+    if (!hasNonUniqueInputs) {
```

Comment:
Using `hasUniqueKeyInAllInputs` might be better for reading (I was confused on ! + `Non` combination). Basically this says  all inputs must have unique key on the same column otherwise return.

Reply:
This variable tracks whether there is at least one input where the group key is not unique. If there are no such inputs (i.e. all inputs already unique on the grouping key), then we return, because the pushdown doesn't make sense. Otherwise - if there is at least one non-unique input - we do the pushdown to all inputs. 

Before the fix, we may do a push-down only to some inputs, which might lead to row type mismatch.

Reply:
Ack. Thanks for the clarification.

Reply:
Makes sense, I've renamed the variable.